text/0020-unique-record-keys.md

```diff
@@ -0,0 +1,103 @@
+
+## Motivation
+
+There are two main motivations for this change.
+
+First, in the status quo, Cedar policy ASTs actually cannot losslessly roundtrip to EST and back.
+This is because the AST representation correctly allows duplicate keys in record literals, but the EST representation drops them (due to an oversight during EST design). In this way, the EST is unfaithful to Cedar semantics, and not all valid Cedar policies are expressible in the EST, which violates the EST's design goal of being lossless modulo whitespace/comments/etc.
+
+More precisely, if a JSON EST contains duplicate record keys in a record literal, our JSON parser will take last-value-wins, and the internal EST representation will contain only the last value.
+This results in the correct behavior on examples like `{ foo: 2, foo: 3 }`, but the incorrect behavior on examples containing errors, like `{ foo: 1 + "hello", foo: 3 }`.
+It also means that roundtripping the AST representation to the EST representation and back necessarily loses information -- specifically, the values associated with all but the last instance of each key.
+
+The second motivation is that the status quo makes things difficult for our Cedar formal specification and accompanying proofs.
+The proofs are required to reason about "well-formed" records as a pre- and post-condition, which could be avoided if records were structurally prohibited from having duplicate keys (e.g., by using a map type instead of an association list).
+The status quo also requires reasoning about the order keys appear in a record literal, meaning that the keys can't be naively sorted without preserving information about their original order.
+If duplicate keys were prohibited, the order of keys wouldn't matter (but see "Unresolved questions" below).
+
+A tertiary motivation for this proposal is that it would bring the Cedar language in line with our desired behavior on duplicate keys in other Cedar API surfaces. See [cedar#159](https://github.com/cedar-policy/cedar/issues/159).
+
+## Detailed design
+
+We'd do something very like [cedar#169](https://github.com/cedar-policy/cedar/pull/169), except that instead of implicitly dropping all but the last value for each key (in `.collect()`), we'd be careful to emit an error when constructing an `Expr` if the record has duplicate keys.
+
+In the case of record literals, the error could be caught in the CST->AST step, resulting in the user seeing a parse error rather than an evaluation error.
+In the case of `context`, the error would need to be caught when constructing the `Context` object.
+In other cases, e.g. if evaluating a Cedar subexpression produces a record with duplicate keys, the user would see an evaluation error. (But see "Unresolved questions" below.)
+
+I don't believe this change would have a significant effect on validation or width/depth subtyping of records, but please correct me if I'm wrong.
```

Comment:
FWIW, for models and proofs, there is no advantage for checking for duplicates in the parser.  To take advantage of this check, the model evaluator would need to carry around complex preconditions that state record literals are free of duplicates.  This is more trouble than checking for duplicates in the evaluator and throwing an error if some are found.

Reply:
@emina's comment above implies that actually we should be able to avoid evaluation errors for duplicate keys, which makes this moot.

Reply:
Sorry, I should have been more clear:  it doesn't help to do this check in the parser, at least not for proofs.  It would be harder to take advantage of that check in the evaluator than to just do the check in the evaluator.

Reply:
Thinking about this again:  the RFC needs to specify (a) the evaluation order, and (b) where duplicate key error is thrown and in what order wrt other errors.

To see why, consider this record:  `{ “one”: 1, “one”:  1 - “three” }`  

What error should this throw?

If the duplicate check is done in the parser, then it should be “Duplicate key error.”

If the duplicate check is done in the evaluator, then there are two options: duplicate error or type error.  Which is thrown depends on _when_ we check for duplicates.  Is it before or after evaluating the RHS of a key-value expression?

Reply:
Updated

Reply:
> it doesn't help to do this check in the parser ... just perform the check in the evaluator.

(1) Performing the check in the parser is better in the sense that it catches the error earlier, eg, when the user constructs a Policy object, long before they get around to calling `is_authorized()`.
(2) Performing the check in the evaluator fails to address the Motivation 1 for this RFC.
(3) Performing the check in the parser amortizes the check across multiple evaluations of the policy.  Performing the check in the evaluator would have to be repeatedly and redundantly performed every time the policy is evaluated.

Reply:
Updated the RFC text to describe error precedence (in Detailed Design)